Source/NHibernate.Extensions/Linq/IncludeQueryProvider.cs

```diff
@@ -327,7 +332,8 @@ private QueryInfo FetchFromPath(QueryInfo queryInfo, string path)
                     queryInfo.CollectionPath = collPath;
                 }
 
-                meta = Session.Factory.GetClassMetadata(assocType.GetAssociatedEntityName(Session.Factory));
+                if (isEntityCollection)
+                    meta = Session.Factory.GetClassMetadata(assocType.GetAssociatedEntityName(Session.Factory));
```

Comment:
the `meta` variable should always be set not only for collections but also for entity relations in order to correctly get the property in the next iteration. Also, I would throw a meaningful exception when the variable is `null` in the next iteration. The code would look like:
```
if (meta == null)
{
    throw new InvalidOperationException($"Unable to fetch property {propName} from path '{path}', no class metadata was found");
}
var propType = meta.GetPropertyType(propName);
if (propType == null)
    throw new Exception($"Property '{propName}' does not exist in the type '{meta.MappedClass.FullName}'");

if (!(propType is IAssociationType assocType))
{
    throw new Exception($"Property '{propName}' does not implement IAssociationType interface");
}

if (assocType.IsCollectionType)
{
    var collectionType = assocType as CollectionType;
    var collectionPersister = (IQueryableCollection)Session.Factory.GetCollectionPersister(collectionType.Role);
    meta = collectionPersister.ElementType.IsEntityType
        ? Session.Factory.GetClassMetadata(collectionPersister.ElementPersister.EntityName)
        : null;

    var collPath = string.Join(".", paths.Take(index + 1));

    //Check if we can fetch the collection without create a cartesian product
    //Fetch can occur only for nested collection
    if (!string.IsNullOrEmpty(queryInfo.CollectionPath) &&
        !collPath.StartsWith(queryInfo.CollectionPath))
    {
        //We have to continue fetching within a new base query
        var nextQueryInfo = queryInfo.GetOrCreateNext();
        return FetchFromPath(nextQueryInfo, path);
    }

    queryInfo.CollectionPath = collPath;
}
else
{
    meta = Session.Factory.GetClassMetadata(assocType.GetAssociatedEntityName(Session.Factory));
}

MethodInfo fetchMethod;

//Try to get the actual property type (so we can skip casting as relinq will throw an exception)
var relatedProp = currentType.GetProperty(propName);

var relatedType = relatedProp != null ? relatedProp.PropertyType : meta?.MappedClass; // Added the ?. in order to prevent a NullReferenceException
```

Reply:
Ok, I changed it to that. But what exactly do you mean with "entity relations"? I'd like to add unit tests for those too.

Reply:
> what exactly do you mean with "entity relations"?

I meant for many to one relations, the following test would fail with the previous code:
```
using (var session = NHConfig.OpenSession())
{
  var query = session.Query<EQBPerson>().Include(o => o.Identity.Owner);
}
```
because the `meta` variable would not be updated and in the next iteration the following line will fail:
```
meta.GetPropertyType(propName);
```
because the `meta` will be still for `EQBPerson` instead for `EQBIdentity` class and it will try to get the `Owner` property from `EQBPerson` which do not exist.